### DISCUSSION TITLE
Providing rationale for why cops exist

### DISCUSSION THREAD

**Discussion by dvandersluis** [rubocop/rubocop#13282] [Ideas]
In https://tenderlovemaking.com/2024/09/29/eliminating-intermediate-array-allocations/, @tenderlove writes:

> I don’t really know RuboCop very well, but I think that in cases like this it would be really helpful if the linter were to tell you _why_ a particular rule is a rule. Personally, I dislike following rules unless I understand the reason behind them. Even if the reasoning is simply “this is just how our team styles our code”. If such a feature is already available in RuboCop, then please feel free to link to this blog post for this particular rule.

I think this is a fair point. The offense messages are usually short and direct ("Do x." "Don't do y.") and if there is a rationalization in the manual it is often not easily evident from actually running RuboCop, particularly for users who are newer to RuboCop or Ruby. It can be disconcerting to have a tool demand / suggest (tone is left up to the reader) you do something that you don't understand why you should want to.

What do people think about making a mode for RuboCop that includes more context about why a cop is reporting? As an idea, perhaps a command line switch to add a secondary one-line rationale below each offense, along with a message printed at the bottom of the offense report when the switch is not enabled along the lines of "Use `--show-rationale` for more context into why RuboCop has reported these offenses" (actual text, switch name, etc. tbd of course!). The cops that report different messages based on context would be an edge case to handle there.

Style cops are going to be the hardest to make improvements on of course, especially since they're the most customizable and generally just boil down to team style / standard.rb style / RuboCop style depending on the codebase.

**Comment by bbatsov** [comment]
No argument from me. I hope it's obvious that the main reason things are the way they are is just lack of time to provide more detailed rationales. (as least as it comes to me) More broadly - nobody expressed a big desire to do the necessary work - documentation is super important, but few people are willing to spend much time improving it. 😅  I've been meaning to add such for the style guide as well for ages, but never got to doing so.

**Comment by bbatsov** [comment]
> As an idea, perhaps a command line switch to add a secondary one-line rationale below each offense, along with a message printed at the bottom of the offense report when the switch is not enabled along the lines of "Use --show-rationale for more context into why RuboCop has reported these offenses" (actual text, switch name, etc. tbd of course!). The cops that report different messages based on context would be an edge case to handle there.

Fine by me. Perhaps this can be some combination from the cop's description comment and a style guide link (if present)?

**Reply by Earlopain** [reply]
I think, most cop documentation will not lend itself easily to being displayed in the terminal since it's not structured enough.

I've recently made a bit of work that a cops documentation can be exposed in editors through extensions like `ruby-lsp`. That previously only worked for RuboCop buildin cops and now extensions can contribute too. https://github.com/rubocop/rubocop/issues/12884

That blue link takes you directly to the docs where you can read description, examples, safety notes, etc. I think linking there instead would make more sense. It's already available and doesn't require much work to implement. Maybe `--show-docs-url` can be enhanced to print the urls during a run for each offense.

If a cops documentation doesn't already contain the rationale in its description (which some likely do) then it should ideally be changed.

**Comment by rubyFeedback** [comment]
I have no real strong personal preference but I agree with Aaron (tenderlove).

Perhaps for each behaviour there could either be an individual .md file explaining it,
or the rationale be maintained in one large .yml file (for instance), along with a time
stamp when it was last updated; and perhaps also an optional author entry. So the
initial version could be written by bbatsov and modified over time; and whenever a
new modification is made, the entry date updated too.

(The reason why it is not that important to me is that we can change the cop behaviour
any way, so any default rule is just as good as any modified rule, since we can automatically
ensure each variant anyway, via the corresponding cop, right? So to me, it is as if different
projects can each use different specifications and rubocop can help enforce these on a
per project / per team basis.)

**Comment by forthrin** [comment]
Been missing URLs (in Emacs) *forever*. Constantly switching to `*Messages*`, copy-paste, search, click, scroll. So definitely supporting this (plus an Emacs keystroke to do `open <url>` will make this pure manna.)

Inlining the full documentation section seems hefty as they are quite thorough at times, including tables and what have you. Guess it *could* be done if the source is already (or can be converted to) Markdown which would fit snugly in the Terminal.

When it comes to what the cop says when you roll down the window, I do support on a general level it's better to say, "You should fix your tail light or someone might run into you" than just saying "You should fix your tail light." A quick stab at some examples:

1. `Useless assignment to variable - out.` → `out is not used anymore, so the whole assignment can be removed.`
1. `Use "gem '#{gem}'" directly instead of format.` → `using format is not necessary when the string has no % codes`
1. `Unused method argument - ai. If it's necessary, use _ or _ai as an argument name to indicate that it won't be used. If it's unnecessary, remove it.` # This one is quite good as it is. Bit long.
